leap/multistep/__init__.py

```diff
@@ -385,35 +424,176 @@ def rk_bootstrap(self, cb):
+            from leap.implicit import generate_solve
+            generate_solve(cb, unknowns, equations, rhs_var_to_unknown, self.state)
+
+        del equations[:]
+        knowns.update(unknowns)
+        unknowns.clear()
+
+        # }}}
+
+        # Update the state now that we've solved.
         if self.state_filter is not None:
             state_est = self.state_filter(state_est)
+        cb(self.state, state_est)
+
+        # Rotate history and time history.
+        self.rotate_and_yield(cb, history, time_history_data)
+
+    def rk_bootstrap(self, cb):
+        """Initialize the timestepper with an IMPLICIT RK method."""
+
+        from leap.rk import IMPLICIT_ORDER_TO_RK_METHOD_BUILDER
+        rk_method = IMPLICIT_ORDER_TO_RK_METHOD_BUILDER[self.function_family.order]
+        rk_coeffs = rk_method.output_coeffs
+        stage_coeff_set_names = ("implicit",)
+        stage_coeff_sets = {"implicit": rk_method.a_implicit}
+        estimate_coeff_set_names = ("main",)
+        estimate_coeff_sets = {"main": rk_coeffs}
+        rhs_funcs = {"implicit": var("<func>"+self.component_id)}
+
+        if self.extra_bootstrap:
+            first_save_step = 2
+        else:
+            first_save_step = 1
+
+        with cb.if_(self.step, "==", first_save_step):
+            # Save the first RHS to the AM history
+            rhs_var = var("rhs_var")
+
+            cb(rhs_var, self.eval_rhs(self.t, self.state))
+            cb(self.history[0], rhs_var)
+
+            if not self.static_dt:
+                cb(self.time_history[0], self.t)
+
+        # Traverse RK stage loop of appropriate order and update state.
+        rk = rk_method(self.component_id, self.state_filter_name)
+        cb = rk.generate_butcher_init(cb, stage_coeff_set_names,
+                                      stage_coeff_sets, rhs_funcs,
+                                      estimate_coeff_set_names,
+                                      estimate_coeff_sets)
+        cb, rhss, est_vars = rk.generate_butcher_primary(cb, stage_coeff_set_names,
+                                                         stage_coeff_sets, rhs_funcs,
+                                                         estimate_coeff_set_names,
+                                                         estimate_coeff_sets)
 
         # Assign the value of the new state.
-        cb(self.state, state_est)
+        cb(self.state, est_vars[0])
+
+        # Save the "next" RHS to the AM history
+        rhs_next_var = var("rhs_next_var")
+
+        cb(rhs_next_var, self.eval_rhs(self.t + self.dt, self.state))
+
+        for i in range(1, len(self.history)):
+            if self.extra_bootstrap:
+                save_crit = i+1
+            else:
+                save_crit = i
+
+            with cb.if_(self.step, "==", save_crit):
+                cb(self.history[i], rhs_next_var)
+
+                if not self.static_dt:
+                    cb(self.time_history[i], self.t + self.dt)
```

Comment:
This code seems like it's an adverse consequence of the awkward redefinition of "history".

Reply:
That and the extra bootstrapping. Given that it's a function of those two things, I'll need to think about how those things need to change before addressing this.

Reply:
This is addressed along with redefinition of history by 4a87ba9.